### PR TITLE
Fix: Windows path seperators break some regexes

### PR DESCRIPTION
**Pull request by Idrinth**
Having tests fail on windows due to insufficiently escaped paths was a bit surprising, so comparisons are now handled with unix-like paths instead and values are quoted where required.

### CODE REVIEW

File: src/Composer/Iterator/SourceFileIteratorFactory.php [hunk 1]
```diff
@@ -156,6 +156,12 @@ public function __construct(
                 string ...$excludePatterns
             ) {
                 if (!empty($excludePatterns)) {
+                    array_walk(
```
Comment (johmanx10) [inline src/Composer/Iterator/SourceFileIteratorFactory.php]: I would prefer a solution that doesn't use references to map an array. I suggest applying `array_map` inline inside the implode of line 167:

```php
$this->excludePattern = sprintf(
    '@^(%s)$@',
    implode(
        '|',
        array_map(
            function (string $pattern): string {
                return preg_quote(
                    str_replace('\\', '/', $pattern),
                    '@'
                );
            },
            $excludePatterns
        )
    )
);
```

Comment (Idrinth) [reply src/Composer/Iterator/SourceFileIteratorFactory.php]: I think the private method would be a better solution here.
Neither the current nor the previous implementation are able to handle glob patterns as allowed by composer, extracting the code into a method would make adjusting that later easier.

**Comment by Idrinth** [comment]
will have a look how to get a windows unittest step in, @johmanx10 , I assume it should be included within scrunitizer?

**Comment by johmanx10** [comment]
@Idrinth I'm not sure Scrutinizer supports Windows. I guess it'll come down to adding a second CI.

**Comment by johmanx10** [comment]
I see that AppVeyor is free for OSS: https://www.appveyor.com/

**Comment by johmanx10** [comment]
@Idrinth I'll be setting up an AppVeyor account and making sure all tests are run on a Windows environment using that. If that all works out, I'll open up a pull request to your branch. Once you accept that, it will start testing your changes. Our goal is to get both CI environments to pass at the same time. If you would then focus on introducing a new private method, I think we're golden.

**Comment by Idrinth** [comment]
thanks @johmanx10, did work on a configuration for appveyor, but I didn't end up with anything working so far.
Both PHPUnit and Composer have a config file that installs php and their respective dependencies, that might be a starting point.

Will have time to do the code change either in my lunch break or tonight after work.

**Comment by johmanx10** [comment]
@Idrinth I opened a [pull request](https://github.com/Idrinth/dependency-guard/pull/1) to your fork, holding the configuration of `appveyor.yml` and the addition of a build badge in `README.md`.

**Comment by Idrinth** [comment]
@johmanx10 I'm unsure about how to do regression tests here, because the build did not succeed originally on windows, meaning the issues were found and covered by existing tests.

**Comment by johmanx10** [comment]
@Idrinth could you show the error you are getting from before you applied a fix? I hope that'll help me think of a good way to write a regression test for this.

**Comment by Idrinth** [comment]
I'm sorry I'm a bit slow with answers. As mentioned before, the issues are already caught, unittest failures before the fixes are:
```
Mediact\DependencyGuard\Tests\Candidate\CandidateExtractorTest::testExtract with data set #3()
Failed asserting that actual size 0 matches expected size 1.
dependency-guard\tests\Candidate\CandidateExtractorTest.php:48

Mediact\DependencyGuard\Tests\Candidate\CandidateExtractorTest::testExtract with data set #4()
Failed asserting that actual size 0 matches expected size 1.
dependency-guard\tests\Candidate\CandidateExtractorTest.php:48

Mediact\DependencyGuard\Tests\Candidate\CandidateExtractorTest::testExtract with data set #5()
Failed asserting that actual size 0 matches expected size 1.
dependency-guard\tests\Candidate\CandidateExtractorTest.php:48

Mediact\DependencyGuard\Tests\Composer\Command\Exporter\TextViolationExporterTest::testExport with data set #2()
preg_replace(): No ending delimiter '#' found
dependency-guard\src\Composer\Command\Exporter\TextViolationExporter.php:64
dependency-guard\src\Composer\Command\Exporter\TextViolationExporter.php:69
dependency-guard\tests\Composer\Command\Exporter\TextViolationExporterTest.php:71

Mediact\DependencyGuard\Tests\Composer\Iterator\SourceFileIteratorFactoryTest::testCreate with data set #4()
preg_match(): Compilation failed: PCRE does not support \L, \l, \N{name}, \U, or \u at offset 5
dependency-guard\src\Composer\Iterator\SourceFileIteratorFactory.php:184
dependency-guard\src\Composer\Iterator\SourceFileIteratorFactory.php:56
dependency-guard\tests\Composer\Iterator\SourceFileIteratorFactoryTest.php:45
```
Those all boil down to mixed directory separators and/or broken regexp due to unescaped pathes being passed in that suddenly contain \ due to the windows path separator.

**Comment by johmanx10** [comment]
@Idrinth I opened a pull request to your fork holding regression tests.

**Comment by Idrinth** [comment]
@johmanx10 haven't found a way around removing the method specific coverageannotation without either having phpunit fail or phpcs besides making themethod cover the class explicitly. The datasets used seem to differ too much to be able to use the same coverage notation otherwise.

**Comment by johmanx10** [comment]
@Idrinth Ah, I see where this went wrong. Because of the narrow diff, I didn't see this before. The new private method you added should have been on the outer class. You can then prepare the pattern before you give it to the constructor of the anonymous class. That should make it all much cleaner and give you the ability to add the annotation in the manner I suggested. Could you please do that?

**Comment by johmanx10** [comment]
@Idrinth I'm really appreciative of your quick response times and invested time and efforts. As a final check, I created a code coverage report against your latest changes. It turns out the code coverage is still down from 100% to now 97%. Some scenarios for the `CandidateExtractor` are no longer covered since your patches were introduced. See:

Candidate namespace

And

extractPackage method

**Comment by Idrinth** [comment]
@johmanx10 will continue to work on that after lunch. Haven't found a reason for coverage to drop in the diff, both tests and the code itself seem to have been unchanged.
When I only run the original test class I end up at 100% coverage, only when including the regression test the coverage suddenly drops.
candidate-extractor
I'd propose running a testsuite that excludes regression tests for coverage, this looks like a phpunit bug at first glance to me.

**Comment by Idrinth** [comment]
just for the record @johmanx10 , I'm just contributing to a tool I'm using and that was build by others. I'm the one who should thank you for all the work that was already put in.

File: phpunit.xml [hunk 2]
```diff
@@ -3,9 +3,12 @@
           bootstrap="vendor/autoload.php">
     <testsuites>
         <testsuite name="default">
-            <directory>tests/Regression</directory>
```
Comment (johmanx10) [inline phpunit.xml]: Please don't remove this line. It is to ensure the regression tests run first. This has got to do with something in the environment I can't quite put my finger on. If the regression tests don't run first, they somehow sometimes pass when they should not.

Comment (Idrinth) [reply phpunit.xml]: restored, sounds as if we might need to change those to be less flaky at some point

**Comment by johmanx10** [comment]
> Edit: would need the ci to run with --testsuite=coverage
- [This comment](https://github.com/mediact/dependency-guard/pull/31#issuecomment-441363052)

- [ ] Update Scrutinizer settings after this pull request is merged to master.

**Pull request merged**